mono/profiler/proflog.c

```diff
@@ -487,8 +488,102 @@ ign_res (int G_GNUC_UNUSED unused, ...)
 {
 }
 
-#define ENTER_LOG(LB, STR)
-#define EXIT_LOG(LB)
+/*
+ * These macros create a scope to avoid leaking the buffer returned
+ * from ensure_logbuf () as it may have been invalidated by a GC
+ * thread during STW.
+ */
+
+#define ENTER_LOG \
+	do { \
+		buffer_lock ()
+
+#define EXIT_LOG \
+		buffer_unlock (); \
+	} while (0)
+
+static volatile gint32 buffer_rwlock_count;
+static volatile gpointer buffer_rwlock_exclusive;
+
+// Can be used recursively.
+static void
+buffer_lock (void)
+{
+	gpointer tid = (gpointer) thread_id ();
+	gpointer excl = InterlockedReadPointer (&buffer_rwlock_exclusive);
+
+	/*
+	 * If the thread holding the exclusive lock tries to modify the
+	 * reader count, just make it a no-op. This way, we also avoid
+	 * invoking the GC safe point macros below, which could break if
+	 * done from a thread that is currently the initiator of STW.
+	 *
+	 * In other words, we rely on the fact that the GC thread takes
+	 * the exclusive lock in the gc_event () callback when the world
+	 * is about to stop.
+	 */
+	if (excl == tid)
+		goto done;
+
+	MONO_ENTER_GC_SAFE;
+
+	while ((excl = InterlockedReadPointer (&buffer_rwlock_exclusive)) && excl != tid)
```

Comment:
It can't, this looks like a leftover from an older (non-working) design of the lock.